From bc5c87afa22f57a7edadbbd1801defd043c3bbfe Mon Sep 17 00:00:00 2001
From: Elizabeth Figura <zfigura@codeweavers.com>
Date: Tue, 21 Oct 2025 13:21:15 -0500
Subject: [PATCH 39/48] ntdll: Handle . and .. segments in relative symlinks.

Note that we don't want to use collapse_path() as-is, because everything else it
does (i.e. forward slashes, duplicate slashes, trailing dots) should explicitly
not apply to symlinks. We also should fail when rewinding past the drive root,
rather than simply ignoring those .. segments.

The FIXME introduced here related to unwinding past the RootDirectory handle is
not particularly problematic to address (we just need to query the server for
its whole path). I've left it as a FIXME simply because it takes extra work to
implement and is unlikely to come up; RootDirectory not used anywhere in
kernel32 or kernelbase file code.

(cherry picked from commit 6b10e8e900ea7910a5f4fe46bb708d0318fe185b)

CW-Bug-Id: #26492
---
 dlls/ntdll/tests/file.c | 12 +++---
 dlls/ntdll/unix/file.c  | 96 ++++++++++++++++++++++++++++++++++++++---
 2 files changed, 97 insertions(+), 11 deletions(-)

diff --git a/dlls/ntdll/tests/file.c b/dlls/ntdll/tests/file.c
index 636706188fe..b3d25c3f139 100644
--- a/dlls/ntdll/tests/file.c
+++ b/dlls/ntdll/tests/file.c
@@ -7123,7 +7123,7 @@ static void test_reparse_points(void)
 
     RtlInitUnicodeString( &nameW, L"testreparse_dirlink\\file" );
     status = NtOpenFile( &handle2, READ_CONTROL, &attr, &io, 0, 0 );
-    todo_wine ok( !status, "got %#lx\n", status );
+    ok( !status, "got %#lx\n", status );
     NtClose( handle2 );
 
     data_size = init_reparse_symlink( &data, L".\\testreparse_dir\\..\\.\\testreparse_dir2\\.", SYMLINK_FLAG_RELATIVE );
@@ -7133,13 +7133,13 @@ static void test_reparse_points(void)
     RtlInitUnicodeString( &nameW, L"testreparse_dirlink" );
     status = NtCreateFile( &handle2, GENERIC_ALL, &attr, &io, NULL, 0, FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE,
                            FILE_OPEN, FILE_DIRECTORY_FILE, NULL, 0 );
-    todo_wine ok( status == STATUS_OBJECT_NAME_NOT_FOUND, "got %#lx\n", status );
+    ok( status == STATUS_OBJECT_NAME_NOT_FOUND, "got %#lx\n", status );
 
     status = NtCreateFile( &handle2, GENERIC_ALL, &attr, &io, NULL, 0, FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE,
                            FILE_OPEN_IF, FILE_DIRECTORY_FILE, NULL, 0 );
-    todo_wine ok( !status, "got %#lx\n", status );
+    ok( !status, "got %#lx\n", status );
     status = NtSetInformationFile( handle2, &io, &fdi, sizeof(fdi), FileDispositionInformation );
-    todo_wine ok( !status, "got %#lx\n", status );
+    ok( !status, "got %#lx\n", status );
     NtClose( handle2 );
 
     data_size = init_reparse_symlink( &data, L"./testreparse_dir", SYMLINK_FLAG_RELATIVE );
@@ -7161,7 +7161,7 @@ static void test_reparse_points(void)
     ok( !status, "got %#lx\n", status );
 
     status = NtOpenFile( &handle2, READ_CONTROL, &attr, &io, 0, 0 );
-    todo_wine ok( status == STATUS_OBJECT_NAME_NOT_FOUND, "got %#lx\n", status );
+    ok( status == STATUS_OBJECT_NAME_NOT_FOUND, "got %#lx\n", status );
 
     data_size = init_reparse_symlink( &data, L"", SYMLINK_FLAG_RELATIVE );
     status = NtFsControlFile( handle, NULL, NULL, NULL, &io, FSCTL_SET_REPARSE_POINT, data, data_size, NULL, 0 );
@@ -7183,7 +7183,7 @@ static void test_reparse_points(void)
 
     RtlInitUnicodeString( &nameW, L"testreparse_dirlink\\" );
     status = NtOpenFile( &handle2, READ_CONTROL, &attr, &io, 0, 0 );
-    todo_wine ok( status == STATUS_IO_REPARSE_DATA_INVALID, "got %#lx\n", status );
+    ok( status == STATUS_IO_REPARSE_DATA_INVALID, "got %#lx\n", status );
 
     /* Create an absolute symlink. */
 
diff --git a/dlls/ntdll/unix/file.c b/dlls/ntdll/unix/file.c
index 0cbc52b7abc..1443e5f1635 100644
--- a/dlls/ntdll/unix/file.c
+++ b/dlls/ntdll/unix/file.c
@@ -4183,6 +4183,76 @@ static NTSTATUS resolve_absolute_reparse_point( const WCHAR *target, unsigned in
 }
 
 
+/* limited version of collapse_path() that only deals with . and .. elements
+ * in relative symlinks */
+static NTSTATUS collapse_relative_symlink( WCHAR *path, unsigned int len, unsigned int *ret_len )
+{
+    const WCHAR *end = path + len;
+    WCHAR *p, *start, *next;
+
+    if (path[0] == '\\')
+    {
+        p = path + 4;
+        while (*p && *p != '\\') p++;
+        p++;
+    }
+    else
+    {
+        p = path;
+    }
+    start = p;
+
+    while (p < end)
+    {
+        if (*p == '.')
+        {
+            if (p + 1 == end) /* final . */
+            {
+                if (p > start) p--;
+                end = p;
+                continue;
+            }
+            else if (p[1] == '\\') /* .\ component */
+            {
+                next = p + 2;
+                memmove( p, next, (end - next) * sizeof(WCHAR) );
+                end -= 2;
+                continue;
+            }
+            else if (p[1] == '.')
+            {
+                if (p + 2 == end) /* final .. */
+                {
+                    if (p == start) return STATUS_IO_REPARSE_DATA_INVALID;
+                    p--;
+                    while (p > start && p[-1] != '\\') p--;
+                    if (p > start) p--;
+                    end = p;
+                    continue;
+                }
+                else if (p[2] == '\\') /* ..\ component */
+                {
+                    if (p == start) return STATUS_IO_REPARSE_DATA_INVALID;
+                    next = p + 3;
+                    p--;
+                    while (p > start && p[-1] != '\\') p--;
+                    memmove( p, next, (end - next) * sizeof(WCHAR) );
+                    end -= (next - p);
+                    continue;
+                }
+            }
+        }
+
+        /* skip to the next component */
+        while (p < end && *p != '\\') p++;
+        if (p < end) p++;
+    }
+
+    *ret_len = end - path;
+    return STATUS_SUCCESS;
+}
+
+
 static NTSTATUS resolve_reparse_point( int fd, int root_fd, OBJECT_ATTRIBUTES *attr, UNICODE_STRING *nt_name,
         unsigned int nt_pos, unsigned int reparse_len, char **unix_name, int unix_len, int pos,
         UINT disposition, BOOL open_reparse, BOOL is_unix, unsigned int reparse_count )
@@ -4228,6 +4298,7 @@ static NTSTATUS resolve_reparse_point( int fd, int root_fd, OBJECT_ATTRIBUTES *a
 
         if (data->SymbolicLinkReparseBuffer.Flags & SYMLINK_FLAG_RELATIVE)
         {
+            unsigned int collapsed_len;
             WCHAR *new_nt_name;
 
             TRACE( "target %s\n", debugstr_wn(target, target_len) );
@@ -4246,17 +4317,32 @@ static NTSTATUS resolve_reparse_point( int fd, int root_fd, OBJECT_ATTRIBUTES *a
 
             memcpy( new_nt_name, name, nt_pos * sizeof(WCHAR) );
             memcpy( new_nt_name + nt_pos, target, target_len * sizeof(WCHAR) );
+
+            if ((status = collapse_relative_symlink( new_nt_name, nt_pos + target_len, &collapsed_len )))
+            {
+                if (attr->RootDirectory)
+                {
+                    /* FIXME: it's legal to unwind past the root directory (but
+                     * not past the volume root), which we can't detect here.
+                     * We need to retrieve the whole NT name */
+                    FIXME( "attempt to unwind past root directory %s\n", debugstr_wn(target, target_len) );
+                }
+                free( new_nt_name );
+                free( data );
+                return status;
+            }
+
             if (remainder_len)
             {
-                if (target[target_len - 1] != '\\')
-                    new_nt_name[nt_pos + target_len++] = '\\';
-                memcpy( new_nt_name + nt_pos + target_len, remainder, remainder_len * sizeof(WCHAR) );
+                if (new_nt_name[collapsed_len - 1] != '\\')
+                    new_nt_name[collapsed_len++] = '\\';
+                memcpy( new_nt_name + collapsed_len, remainder, remainder_len * sizeof(WCHAR) );
             }
-            new_nt_name[nt_pos + target_len + remainder_len] = 0;
+            new_nt_name[collapsed_len + remainder_len] = 0;
 
             free( nt_name->Buffer );
             nt_name->Buffer = new_nt_name;
-            nt_name->Length = (nt_pos + target_len + remainder_len) * sizeof(WCHAR);
+            nt_name->Length = (collapsed_len + remainder_len) * sizeof(WCHAR);
             nt_name->MaximumLength = nt_name->Length + sizeof(WCHAR);
             attr->ObjectName = nt_name;
 
-- 
2.53.0

